### PR TITLE
Simplify Stylus

### PR DESCRIPTION
**Pull request by TheAlphaNerd**
This uses npm scripts rather than grunt to do the stylus build.  Removes quite a few dependencies and scripts to maintain.

### CODE REVIEW

File: public/styl/global.styl [hunk 1]
```diff
@@ -531,4 +531,4 @@ article .posted {
   input, textarea {
     width: 95%;
   }
-}
\ No newline at end of file
+}
```
Comment (isaacs) [inline public/styl/global.styl]: vim ftmfw.

**Comment by isaacs** [comment]
@TheAlphaNerd Hm, maybe I'm doing something wrong?  When I merge this into master, and then run `rm -f ./public/css/main.min.css && npm start`, it creates a file with these three lines at the top:

```
  [90mwatching[0m /Users/isaacs/dev/js/bigboringsystem/node_modules/stylus/lib/functions/index.styl
  [90mwatching[0m public/styl/reset.styl
  [90mwatching[0m public/styl/global.styl
```

It looks like stylus is dropping some informational output in there.  Can that be suppressed?  It's showing up on stdout, not stderr, so it feels like a bug to me.

**Comment by TheAlphaNerd** [comment]
Looking into this right now... great eye @isaacs

**Comment by TheAlphaNerd** [comment]
@isaacs this indeed looks like a weird thing happening with stylus.  They give a weird interface where you can specific output dir, but not filename (afaik).  I'm using a flag to have it print everything to std instead of writing a file and then piping that... it seems that when the watch flag is thrown things get murky.

I'll take a look at this later tonight.

**Comment by thefoxis** [comment]
Hmm, I haven't seen that with compiling with Stylus before. Maybe something got changed in newer version, but I remember using [a similar setup](https://github.com/Level/leveldb.org/blob/master/package.json#L7-L11) in leveldb.org and it worked a-okay.

**Comment by TheAlphaNerd** [comment]
The major difference at leveldb is that they are using the ```-o``` option, which can only specify directory, not file..

Since we want to include min in the filename, we can't use that interface, which is why I am using ```-p```to print to std and pipe the contents to a file.  This is a total hack.

In truth stylus is messing up two things here.  First there should be an interface to output to a specific file name.  Secondly when printing the output while watching, the watch output should definitely not end up in std.out, or it messes up builds.

I'm going to take a stab at making a fix for the problems with the stylus CLI and send a PR.  We should hold off on merging this until there is a cut of stylus that includes in, I do not think it is a good idea to start tracking a fork off github.

**Comment by TheAlphaNerd** [comment]
Fixed by https://github.com/LearnBoost/stylus/pull/1796

I definitely took an easy road here to stop us from getting blocked.  This does not yet deal with specifically naming output.

**Comment by thefoxis** [comment]
@TheAlphaNerd&mdash;sure, I'm not a pro at CLIs and npm at all, was only throwing in an example. If that can be fixed directly in Stylus and incorporated here I think that would be great. Also in the meantime we could get away with not having `min` in the filename and have another PR to fix that. Whatever you think is easier/faster :)

**Comment by TheAlphaNerd** [comment]
@thefoxis I actually really appreciated having that example, it helped me figure out that my hunch wasn't wrong.

I think that it would be best to keep the gulp for the time being until the changes land in stylus.

**Comment by isaacs** [comment]
Wait a minute, though...

If it's watching, and running in the background, and dumping out a whole new version of the CSS every time the stylus changes, won't you end up with a CSS file that's got a whole bunch of copies in it?

It seems like we need to either change stylus to support outputting a specific filename, or accept that it's not called `main.min.css`, or just keep using gulp because why not.

**Comment by TheAlphaNerd** [comment]
Bah... I'm digging more into this right now and it appears that ```---watch```` and ```-p``` just do not work together at all.  : (

@isaacs it would appear your concern is never met as stylus never even properly watches the files when the ```-p``` tag is used.

I created an issue https://github.com/LearnBoost/stylus/issues/1797

**Comment by isaacs** [comment]
@TheAlphaNerd Well, what I'm saying is that `--watch` and `-p` are fundamentally at odds.  Think it through.

You set up a process that watches `main.styl` and every time it changes, it prints the updated CSS.

Then you pipe this watching/printing process to `main.css`.

So, you make a change to `main.styl`.  It outputs some stuff on stdout, and fills up `main.css` with 100 lines of CSS.  `main.css` is now 100 lines.

Then, `main.styl` changes again.  The same watcher process notes the change builds the CSS *again*, and prints it to stdout *again*.

Stdout of the watcher process is piped to `main.css`, so the new 100 lines are appended to the end of `main.css`, which is now 200 lines long.

Make a third change, and `main.css` is now 300 lines long.  Delete half the styles, save the file.  Now the file is 350 lines long, *and the deleted styles are still there!*

The designer, who is editing in their editor and then refreshing in their browser, is confused by two things:

1. The CSS line numbers in the developer tools keep increasing.
2. The page gets slower and slower to load.
3. **When they delete styles, they don't go away.**

A watcher must always create a brand new file to dump the styles into, opening as `w+`.  If you are printing the output to stdout, and piping the stdout to a file, *and doing this every time the file changes in a single process without restarting and truncating the file*, then you're entering a World of Pain.

The bug is that stylus lets you do this at all.

**Comment by TheAlphaNerd** [comment]
I guess I was making an incorrect assumption about the way the piping would work in the npm script.  I was imagining that each time that output would come the watch task it would make a new pipe... this is obviously not the case.

At the current moment I am trying to get the stylus cli to throw if both args are provided.

Next up will be getting in functionality to simply write to a file path.

I have offered to do a refactor / rewrite of their cli + tests if they are open to it.

**Comment by TheAlphaNerd** [comment]
Landed 

https://github.com/LearnBoost/stylus/pull/1798

**Comment by TheAlphaNerd** [comment]
Since I obviously was quite mistaken in the way I was handling the watch task I have rebased and implemented the watch task writing to a file without ```.min``` in the name.

Please do not merge this, I simply want the current implementation of this branch to be working.  i would like to see some changes to the stylus CLI before we rely on it as a build tool.  In the mean time the gulp task is by far the best approach.

**Comment by TheAlphaNerd** [comment]
Got it

https://github.com/LearnBoost/stylus/pull/1800

If this lands it should be in production with a cut they are planning for next week

**Comment by ednapiranha** [comment]
needs rebase

**Comment by TheAlphaNerd** [comment]
This has been rebased and is now working exactly as expected.  We need to wait for a new version of stylus to be cut before merging

**Comment by TheAlphaNerd** [comment]
@ednapiranha @thefoxis @isaacs @llkats 

There was a recent publication to gulp-nodemon that breaks the entire dev stack.

I know that we wanted to wait for stylus to publish the update before merging this, but I think that this breaking change warrants an early merge.  I have tracked stylus in the package.json against the github version, it is tracking a specific commit rather than master (I have verified this version works)

**Comment by rtorr** [comment]
@TheAlphaNerd Fully support this. We want to get off gulp.

**Comment by jasonrhodes** [comment]
Just curious, @rtorr, why? Overkill for this project?

**Comment by TheAlphaNerd** [comment]
Absolutely overkill for this project.  The gulp stack has 4 dependencies and 20+ lines of scripts for what is being accomplished here in one line of package.json

**Comment by jasonrhodes** [comment]
Oh yeah @TheAlphaNerd, makes sense to me—I haven't followed along with this PR at all so I'm not in a position to merge but I don't really care about gulp whatever. 

I was mostly interested in @rtorr's comment because I thought he had introduced gulp a few days ago. I may be mistaken on that memory.

**Comment by isaacs** [comment]
It sounds like the group is settled on removing gulp.  Totally fine with me.

What's this mean for watching with stylus?  Should we just accept that we can't set the filename like we want?  Seems like a pretty minor concession.

**Comment by rtorr** [comment]
@jasonrhodes
> I was mostly interested in @rtorr's comment because I thought he had introduced gulp a few days ago. I may be mistaken on that memory.

It was a means to get some things done quickly, and gulp was an option that was discussed in chat. I don't mind removing things that I introduce.

**Comment by TheAlphaNerd** [comment]
@isaacs watching stylus now works fine using the stylus cli.  I got the pr merged.  If you peek the ```package.json``` you will notice we are installing a specific sha that includes the patch.

**Comment by TheAlphaNerd** [comment]
Awe shucks. Thanks for the sunshine!

**Comment by ednapiranha** [comment]
ruh roh. @TheAlphaNerd rebase?

**Comment by TheAlphaNerd** [comment]
All done did it

**Comment by TheAlphaNerd** [comment]
:+1:

**Pull request merged**